subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyFactory.java

```diff
@@ -0,0 +1,105 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts.dsl;
+
+import org.gradle.api.NonExtensible;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.ExternalModuleDependency;
+import org.gradle.api.artifacts.FileCollectionDependency;
+import org.gradle.api.artifacts.MinimalExternalModuleDependency;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderConvertible;
+import org.gradle.internal.HasInternalProtocol;
+
+import java.util.Map;
+
+/**
+ * Factory class for creating {@link Dependency} instances, with strong typing.
+ *
+ * <p>
+ * <b>Note:</b> This interface is not intended for implementation by build script or plugin authors.
+ * </p>
+ */
+@HasInternalProtocol
+@NonExtensible
+public interface DependencyFactory {
+    ExternalModuleDependency fromCharSequence(CharSequence dependencyNotation);
+
+    default DependencyProvider<ExternalModuleDependency> fromCharSequence(Provider<? extends CharSequence> dependencyNotation) {
+        return fromDependency(dependencyNotation.map(this::fromCharSequence));
+    }
+
+    default DependencyProvider<ExternalModuleDependency> fromCharSequence(ProviderConvertible<? extends CharSequence> dependencyNotation) {
+        return fromCharSequence(dependencyNotation.asProvider());
+    }
+
+    ExternalModuleDependency fromMinimal(MinimalExternalModuleDependency dependencyNotation);
+
+    default DependencyProvider<ExternalModuleDependency> fromMinimal(Provider<? extends MinimalExternalModuleDependency> dependencyNotation) {
+        return fromDependency(dependencyNotation.map(this::fromMinimal));
+    }
+
+    default DependencyProvider<ExternalModuleDependency> fromMinimal(ProviderConvertible<? extends MinimalExternalModuleDependency> dependencyNotation) {
+        return fromMinimal(dependencyNotation.asProvider());
+    }
+
+    ExternalModuleDependency fromMap(Map<String, ?> dependencyNotation);
+
+    default DependencyProvider<ExternalModuleDependency> fromMap(Provider<? extends Map<String, ?>> dependencyNotation) {
+        return fromDependency(dependencyNotation.map(this::fromMap));
+    }
+
+    default DependencyProvider<ExternalModuleDependency> fromMap(ProviderConvertible<? extends Map<String, ?>> dependencyNotation) {
+        return fromMap(dependencyNotation.asProvider());
+    }
```

Comment:
I don't know if that would be very useful for a `Map`? Are there any use cases for that? If not, I'd like to keep it simple, and we can expand it later.

Reply:
I was thinking of consistency with the `fromCharSequence` methods.

Reply:
I was thinking of consistency with the `fromCharSequence` methods.

Reply:
Yes, but the difference there is that `CharSequence`s are frequently used on their own, but I don't know why they would be used for a `Map`, and I would need to convert the `Map` internally to actually use it, because `CharSequence` does not define equality. I don't think anyone should need this in most cases, whereas custom `CharSequence`s for building items are common (e.g. `StringBuilder`).